third_party/terraform/utils/batcher.go

```diff
@@ -0,0 +1,204 @@
+package google
+
+import (
+	"fmt"
+	"log"
+	"sync"
+	"time"
+)
+
+const defaultBatchSendIntervalSec = 10
+
+// RequestBatcher a global batcher object.
+type RequestBatcher struct {
+	sync.Mutex
+
+	*batchingConfig
+	batches map[string]*startedBatch
+}
+
+// BatchRequest represents a single request to a global batcher.
+type BatchRequest struct {
+	// ResourceName represents the underlying resource for which
+	// a request is made. Its format is determined by what SendF expects, but
+	// typically should be the name of the parent GCP resource being changed.
+	ResourceName string
+
+	// Body is this request's data to be passed to SendF, and may be combined
+	// with other bodies using CombineF.
+	Body interface{}
+
+	// CombineF function determines how to combine bodies from two batches.
+	CombineF batcherCombineFunc
+
+	// SendF function determines how to actually send a batched request to a
+	// third party service. The arguments given to this function are
+	// (ResourceName, Body) where Body may have been combined with other request
+	// Bodies.
+	SendF batcherSendFunc
+
+	// ID for debugging request. This should be specific to a single request
+	// (i.e. per Terraform resource)
+	DebugId string
+}
+
+// These types are meant to be the public interface to batchers. They define
+// logic to manage batch data type and behavior, and require service-specific
+// implementations per type of request per service.
+// Function type for combine existing batches and additional batch data
+type batcherCombineFunc func(body interface{}, toAdd interface{}) (interface{}, error)
+
+// Function type for sending a batch request
+type batcherSendFunc func(resourceName string, body interface{}) (interface{}, error)
+
+// batchResponse bundles an API response (data, error) tuple.
+type batchResponse struct {
+	body interface{}
+	err  error
+}
+
+// startedBatch refers to a processed batch whose timer to send the request has
+// already been started. The responses for the request is sent to each listener
+// channel, representing parallel callers that are waiting on requests
+// combined into this batch.
+type startedBatch struct {
+	*BatchRequest
+
+	listeners []chan batchResponse
+	timer     *time.Timer
+}
+
+// batchingConfig contains user configuration for controlling batch requests.
+type batchingConfig struct {
+	sendAfter       time.Duration
+	disableBatching bool
+}
+
+// Initializes a new batcher
+func NewRequestBatcher(config *batchingConfig) *RequestBatcher {
+	return &RequestBatcher{
+		batchingConfig: config,
+		batches:        make(map[string]*startedBatch),
+	}
+}
+
+// SendRequestWithTimeout is expected to be called per parallel call.
+// It manages waiting on the result of a batch request.
+//
+// Batch requests are grouped by the given batchKey, and thus batchKey
+// should be unique to the type of batch request being sent and the affected
```

Comment:
Could this be rephrased to say it be unique to the URL used for the operation? eg: this could be resource_name + project_id?
Armed with this comment I am not sure exactly how to format a given key.

Reply:
@chrisst tried to elaborate - I am still conflicted about how I did this, because it's pretty arbitrary right now without much validation (or at least something like a registry). I think any solution I come up with still has some sort of list of request types per batcher/service and will build off some list of consts/"enum" at the top of each batcher file but it still feels bad :/